### PR TITLE
SCT-124: Use a new func instead of same instance

### PR DESCRIPTION
**Pull request by rikonor**
**Why is this necessary?**

The old function `NewBackOffClient(client Client, backoff Backoff)` creates a client that uses the same `backoff` instance for all requests. This is not thread-safe since a request finishing and resetting the backoff could affect another request that's still running because they share backoff instance.

By making each request have it's own instance of a backoff, this issue is avoided.

### CODE REVIEW

File: backoff/backoff_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	// Fail a few times so backoff will be capped
+	for i := 0; i < 10; i++ {
+		bo.Failure()
+	}
+
+	// Backoff should now be capped
+	if bo.current != maxBackoff {
+		t.Fatalf("expected backoff to be capped at %v but got %v", maxBackoff, bo.current)
+	}
+}
```
Comment (Seanstoppable) [inline backoff/backoff_test.go]: Also need a test for just successes, as well as a bunch of successes and failures.

File: backoff/client.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+type clientWithBackoff struct {
+	mu *sync.Mutex
+
+	client          HTTPClient
+	backoff         Backoff
+	nextRequestTime time.Time
+
+	// maxElapsed is the maximum amount of time per request
+	maxElapsed time.Duration
```
Comment (Seanstoppable) [inline backoff/client.go]: Should we be expiring based on time or based on number of requests?

File: backoff/backoff_test.go (hunk 1 above)
Comment (rikonor) [reply backoff/backoff_test.go]: What are you looking to test? I don't mind adding tests but I don't want redundant ones. I feel proper behavior on success and failure is tested in the first test.

File: backoff/client.go (hunk 2 above)
Comment (rikonor) [reply backoff/client.go]: Can you elaborate?

**Comment by lattwood** [comment]
Still not a big fan of this repository's existence, simply because we can use a mocked `RoundTripper` for tests rather than this.

That opinion will negatively colour any review I do of the code.

**Comment by rikonor** [comment]
@lattwood this change is due to an issue you highlighted, so I think it's reasonable to expect your review.

Can what this repo does be done in other ways? Sure. Isn't that true for pretty much anything?

Do you have a **concrete** reason why this repo should be re-written from scratch using the method you prefer? If not, let's keep things as they are since they seem to serve us well enough.

**Comment by lattwood** [comment]
Concrete reason: the golang authors considered how to middleware an http request and RoundTripper is the way. The decision to not have http.Client behind an interface in the stdlib was a conscious decision by people smarter than us and we should be mindful of that.

**Comment by rikonor** [comment]
That's not a concrete reason. A concrete reason to me would be: "this pattern leads to sub-optimal performance" or "You could run into this issue because of this", backed by evidence of course.

File: backoff/backoff.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		maxBackoff: opts.MaxBackoff,
+	}
+}
+
+func (bo *exponentialBackoff) Success() time.Duration {
+	bo.mu.Lock()
+	defer bo.mu.Unlock()
+
+	// Decrease the current backoff
+	bo.current = time.Duration(float64(bo.current) / bo.multiplier)
```
Comment (Seanstoppable) [inline backoff/backoff.go]: If you have one backoff, there is a bug here where we will continuously get shorter and shorter backoff fractions for multiple successes. This means that when we then need to backoff, we do it very very slowly (not actually doing what we want).

File: backoff/client.go (hunk 2 above)
Comment (Seanstoppable) [reply backoff/client.go]: We don't want to retry forever.

File: backoff/backoff.go (hunk 3 above)
Comment (rikonor) [reply backoff/backoff.go]: I'm not sure I agree this is a bug, but rather the intended behavior all along. We had discussed the application as a whole being throttled upon error, so it makes sense that if a request succeeds somewhere, then the backoff should decrease.

**Comment by rikonor** [comment]
> I think we may be biting off more than we want to here.
Rather than trying to implement our own backoff logic, I would rather go back to the initial 'one backoff multiplier per client'

I'm not sure I agree with that because if we continue to consider a backoff instance to be not thread-safe, then we have to start treating http clients as non-thread-safe.

File: backoff/client.go (hunk 2 above)
Comment (rikonor) [reply backoff/client.go]: I don't follow... `maxElapsed` limits the time a request can be be tried. So if `maxElapsed` is 5 minutes, you will be retrying at most 5 minutes.

File: backoff/backoff.go (hunk 3 above)
Comment (Seanstoppable) [reply backoff/backoff.go]: Sure, the backoff should decrease, but it is not expected for backoff to be 0.000000001ms and then a failure to move it to 0.000000015 ms.

Comment (rikonor) [reply backoff/backoff.go]: Ah now I understand what you meant. Fixing.

Comment (rikonor) [reply backoff/backoff.go]: This is now fixed via https://github.com/securityscorecard/go-httpmock/commit/ecfa066541992d48a514c76232b3c166efaac894

**Pull request merged**